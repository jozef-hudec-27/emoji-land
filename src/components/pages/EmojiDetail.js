import { useParams } from 'react-router-dom'
import { findEmoji } from '../../helpers/emojis'
import AddToCartBtn from '../AddToCartBtn'
import EmojiNotFound from '../emoji/EmojiNotFound'
import ProductCounter from '../ProductCounter'

export default function EmojiDetail() {
  const params = useParams()
  const emojiObj = findEmoji(params.emoji)

  if (!emojiObj)
    return (
      <div className="full-width full-height flexbox flex-justify-center">
        <EmojiNotFound />
      </div>
    )

  const { name, emoji, description } = emojiObj

  return (
    <div className="flexbox flex-justify-center full-height">
      <div className="emoji-detail">
        <p style={{ fontSize: '5rem' }}>{emoji}</p>

        <div className="flexbox flex-align-center full-width">
          <h2>{name}</h2>
          <h3 className="ml-auto" style={{ color: 'green' }}>
            5$
          </h3>
        </div>

        <p style={{ textAlign: 'justify' }}>{description}</p>

        <div className="add-to-cart-container flexbox full-width mt-12">
          <AddToCartBtn />
          <ProductCounter />
        </div>
      </div>
    </div>
  )
}
