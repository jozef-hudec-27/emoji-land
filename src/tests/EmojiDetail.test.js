import { render, screen } from '@testing-library/react'
import EmojiDetail from '../components/pages/EmojiDetail'
import userEvent from '@testing-library/user-event'

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: () => ({ emoji: 'ðŸ˜€' }),
}))

describe('EmojiDetail component', () => {
  it('renders emoji, name, description and price', () => {
    const { container } = render(
      <EmojiDetail
        changeQuantity={jest.fn()}
        addToCart={jest.fn()}
        removeFromCart={jest.fn()}
        cartContains={jest.fn()}
        cartCount={jest.fn()}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('adds emojis to cart after clicking the button', () => {
    const addToCart = jest.fn((emoji) => emoji)

    render(
      <EmojiDetail
        changeQuantity={jest.fn()}
        addToCart={addToCart}
        removeFromCart={jest.fn()}
        cartContains={jest.fn()}
        cartCount={jest.fn()}
      />
    )

    userEvent.click(screen.getByRole('button', { name: 'Add to cart' }))

    expect(addToCart).toHaveBeenCalled()
  })

  it('increases quantity after clicking the button', () => {
    const cartContains = jest.fn()
    cartContains.mockReturnValueOnce(false).mockReturnValueOnce(false).mockReturnValueOnce(true)

    const addToCart = jest.fn()

    const changeQuantity = jest.fn()

    render(
      <EmojiDetail
        changeQuantity={changeQuantity}
        addToCart={addToCart}
        removeFromCart={jest.fn()}
        cartContains={cartContains}
        cartCount={jest.fn()}
      />
    )

    const incrementBtn = screen.getByRole('button', { name: 'Increment' })
    userEvent.click(incrementBtn)
    userEvent.click(incrementBtn)

    expect(addToCart).toHaveBeenCalled()

    expect(changeQuantity).toHaveBeenCalled()
    expect(cartContains).toHaveBeenCalledTimes(3)
  })

  it('decreases the quantity after clicking the button', () => {
    const cartContains = jest.fn()
    cartContains.mockReturnValueOnce(true).mockReturnValueOnce(true).mockReturnValueOnce(false)
    const changeQuantity = jest.fn()

    render(
      <EmojiDetail
        changeQuantity={changeQuantity}
        addToCart={jest.fn()}
        removeFromCart={jest.fn()}
        cartContains={cartContains}
        cartCount={jest.fn()}
      />
    )

    const decrementBtn = screen.getByRole('button', { name: 'Decrement' })
    userEvent.click(decrementBtn)
    userEvent.click(decrementBtn)

    expect(changeQuantity).toHaveBeenCalled()
    expect(cartContains).toHaveBeenCalledTimes(3)
  })
})
