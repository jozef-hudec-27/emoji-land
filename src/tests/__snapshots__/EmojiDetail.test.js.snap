// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmojiDetail component renders emoji, name, description and price 1`] = `
<div>
  <div
    class="flexbox flex-justify-center full-height p-8"
  >
    <div
      class="emoji-detail card flexbox flex-column flex-center"
    >
      <p
        style="font-size: 5rem;"
      >
        ðŸ˜€
      </p>
      <div
        class="flexbox flex-align-center full-width"
      >
        <h2>
          Grinning Face
        </h2>
        <h3
          class="ml-auto"
          style="color: green;"
        >
          $
          7
        </h3>
      </div>
      <p
        style="text-align: justify;"
      >
        A yellow face with simple, open eyes and a broad, open smile, showing upper teeth and tongue on some platforms. Often conveys general pleasure and good cheer or humor.
      </p>
      <div
        class="add-to-cart-container flexbox full-width mt-12"
      >
        <button
          class="add-to-cart"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cart-shopping "
            data-icon="cart-shopping"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 24C0 10.7 10.7 0 24 0H69.5c22 0 41.5 12.8 50.6 32h411c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3H170.7l5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H199.7c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5H24C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
              fill="currentColor"
            />
          </svg>
           Add to cart
        </button>
        <div
          class="product-counter flexbox flex-align-center ml-auto gap-10"
        >
          <button
            aria-label="Increment"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-up "
              data-icon="chevron-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M201.4 105.4c12.5-12.5 32.8-12.5 45.3 0l192 192c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 173.3 54.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l192-192z"
                fill="currentColor"
              />
            </svg>
          </button>
          <p />
          <button
            aria-label="Decrement"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
